In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import plotly.express as px
import seaborn as sns

In [2]:
df=pd.read_csv(r'C:\Users\user\Desktop\Analyst\week 8\WDI_CSV\WDICSV.csv')

In [3]:
df.head()

,Country Name,Country Code,Indicator Name,Indicator Code,1960,1961,1962,1963,1964,1965,...,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
0,Africa Eastern and Southern,AFE,Access to clean fuels and technologies for coo...,EG.CFT.ACCS.ZS,NaN,NaN,NaN,NaN,NaN,NaN,...,18.685118,19.205632,19.742772,20.332679,20.862800,21.419621,21.996456,22.541440,NaN,NaN
1,Africa Eastern and Southern,AFE,Access to clean fuels and technologies for coo...,EG.CFT.ACCS.RU.ZS,NaN,NaN,NaN,NaN,NaN,NaN,...,7.606712,7.926604,8.309896,8.704591,9.106640,9.480804,9.903209,10.288154,NaN,NaN
2,Africa Eastern and Southern,AFE,Access to clean fuels and technologies for coo...,EG.CFT.ACCS.UR.ZS,NaN,NaN,NaN,NaN,NaN,NaN,...,39.052626,39.321068,39.649534,39.968299,40.354628,40.723805,41.026351,41.289974,NaN,NaN
3,Africa Eastern and Southern,AFE,Access to electricity (% of population),EG.ELC.ACCS.ZS,NaN,NaN,NaN,NaN,NaN,NaN,...,38.859598,40.223744,43.035073,44.390861,46.282371,48.127211,48.742043,50.667516,52.598171,NaN
4,Africa Eastern and Southern,AFE,"Access to electricity, rural (% of rural popul...",EG.ELC.ACCS.RU.ZS,NaN,NaN,NaN,NaN,NaN,NaN,...,24.584906,25.375037,26.941984,28.983183,30.909991,32.709837,33.998179,35.240236,36.080572,NaN


In [4]:
df.shape

(396970, 70)

In [5]:
indicators=df.groupby('Indicator Code')['Indicator Name'].value_counts().reset_index()
indicators

,Indicator Code,Indicator Name,count
0,AG.CON.FERT.PT.ZS,Fertilizer consumption (% of fertilizer produc...,265
1,AG.CON.FERT.ZS,Fertilizer consumption (kilograms per hectare ...,265
2,AG.LND.AGRI.K2,Agricultural land (sq. km),265
3,AG.LND.AGRI.ZS,Agricultural land (% of land area),265
4,AG.LND.ARBL.HA,Arable land (hectares),265
...,...,...,...
1493,per_si_allsi.cov_q1_tot,Coverage of social insurance programs in poore...,265
1494,per_si_allsi.cov_q2_tot,Coverage of social insurance programs in 2nd q...,265
1495,per_si_allsi.cov_q3_tot,Coverage of social insurance programs in 3rd q...,265
1496,per_si_allsi.cov_q4_tot,Coverage of social insurance programs in 4th q...,265


In [6]:
indicators.to_csv('Indicator Code.csv',index=False)

                                                               Service-Led Growth vs. Goods Trade                                                                                 

 * Has the expansion of global transport, travel, and financial service trade outperformed physical merchandise trade growth from 2000 to 2024

Key Indicator Codes:

    * BX.GSR.NFSV.CD – Service exports (BoP, current US$)

    * BX.GSR.MRCH.CD – Goods exports (BoP, current US$)

    * BX.GSR.TRAN.ZS – Transport services (% of service exports)

    * BX.GSR.TRVL.ZS – Travel services (% of service exports)

    * BX.GSR.INSF.ZS – Insurance and financial services (% of service exports)

In [7]:
df=df[df['Indicator Code'].isin(['BX.GSR.NFSV.CD', 'BX.GSR.MRCH.CD','BX.GSR.TRAN.ZS','BX.GSR.TRVL.ZS','BX.GSR.INSF.ZS'])]
df.head()

,Country Name,Country Code,Indicator Name,Indicator Code,1960,1961,1962,1963,1964,1965,...,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
553,Africa Eastern and Southern,AFE,"Goods exports (BoP, current US$)",BX.GSR.MRCH.CD,NaN,NaN,NaN,NaN,NaN,NaN,...,1.712346e+11,1.954440e+11,2.172074e+11,2.028628e+11,1.824270e+11,2.581108e+11,2.922997e+11,2.664837e+11,2.790439e+11,2.786310e+11
680,Africa Eastern and Southern,AFE,Insurance and financial services (% of service...,BX.GSR.INSF.ZS,NaN,NaN,NaN,NaN,NaN,NaN,...,5.448314e+00,5.693671e+00,6.024173e+00,5.100350e+00,7.102548e+00,7.309102e+00,5.583394e+00,5.684983e+00,5.092709e+00,9.318626e+00
1311,Africa Eastern and Southern,AFE,"Service exports (BoP, current US$)",BX.GSR.NFSV.CD,NaN,NaN,NaN,NaN,NaN,NaN,...,4.007407e+10,4.444953e+10,4.883753e+10,4.799508e+10,2.972230e+10,3.573699e+10,4.927544e+10,5.332154e+10,5.917083e+10,NaN
1426,Africa Eastern and Southern,AFE,"Transport services (% of service exports, BoP)",BX.GSR.TRAN.ZS,NaN,NaN,NaN,NaN,NaN,NaN,...,2.246033e+01,2.357277e+01,2.362018e+01,2.556941e+01,3.012808e+01,3.021443e+01,2.944156e+01,2.827607e+01,2.747928e+01,1.441637e+01
1430,Africa Eastern and Southern,AFE,"Travel services (% of service exports, BoP)",BX.GSR.TRVL.ZS,NaN,NaN,NaN,NaN,NaN,NaN,...,4.823710e+01,4.720931e+01,4.689255e+01,4.605773e+01,2.897217e+01,2.863691e+01,3.736879e+01,4.065871e+01,4.178790e+01,3.793399e+01


In [8]:
df.info()

<class 'pandas.DataFrame'>
Index: 1325 entries, 553 to 396902
Data columns (total 70 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country Name    1325 non-null   str    
 1   Country Code    1325 non-null   str    
 2   Indicator Name  1325 non-null   str    
 3   Indicator Code  1325 non-null   str    
 4   1960            14 non-null     float64
 5   1961            14 non-null     float64
 6   1962            9 non-null      float64
 7   1963            9 non-null      float64
 8   1964            9 non-null      float64
 9   1965            14 non-null     float64
 10  1966            14 non-null     float64
 11  1967            24 non-null     float64
 12  1968            34 non-null     float64
 13  1969            34 non-null     float64
 14  1970            63 non-null     float64
 15  1971            90 non-null     float64
 16  1972            99 non-null     float64
 17  1973            98 non-null     float64
 18  

In [9]:
df.isnull().sum().iloc[4:]

1960    1311
1961    1311
1962    1316
1963    1316
1964    1316
        ... 
2021     217
2022     218
2023     220
2024     288
2025     699
Length: 66, dtype: int64

In [10]:
year_cols = [str(year) for year in range(2000, 2025)]

missing_summary = (
    df.groupby(['Indicator Code', 'Indicator Name'])[year_cols]
      .apply(lambda x: x.isna().mean().mean() * 100)
      .reset_index(name='Missing_Percentage')
)

missing_summary

,Indicator Code,Indicator Name,Missing_Percentage
0,BX.GSR.INSF.ZS,Insurance and financial services (% of service...,22.913208
1,BX.GSR.MRCH.CD,"Goods exports (BoP, current US$)",16.618868
2,BX.GSR.NFSV.CD,"Service exports (BoP, current US$)",16.377358
3,BX.GSR.TRAN.ZS,"Transport services (% of service exports, BoP)",18.490566
4,BX.GSR.TRVL.ZS,"Travel services (% of service exports, BoP)",18.400000


In [11]:
year_cols = [str(year) for year in range(2000, 2025)]

df = df[
    ['Country Name', 'Country Code',
     'Indicator Name', 'Indicator Code'] + year_cols
]

In [12]:
df.head()

,Country Name,Country Code,Indicator Name,Indicator Code,2000,2001,2002,2003,2004,2005,...,2015,2016,2017,2018,2019,2020,2021,2022,2023,2024
553,Africa Eastern and Southern,AFE,"Goods exports (BoP, current US$)",BX.GSR.MRCH.CD,5.769241e+10,5.545812e+10,5.921214e+10,7.119757e+10,9.052145e+10,1.151804e+11,...,1.807261e+11,1.712346e+11,1.954440e+11,2.172074e+11,2.028628e+11,1.824270e+11,2.581108e+11,2.922997e+11,2.664837e+11,2.790439e+11
680,Africa Eastern and Southern,AFE,Insurance and financial services (% of service...,BX.GSR.INSF.ZS,6.021941e+00,4.334388e+00,3.358394e+00,3.854574e+00,4.324752e+00,4.708117e+00,...,5.359036e+00,5.448314e+00,5.693671e+00,6.024173e+00,5.100350e+00,7.102548e+00,7.309102e+00,5.583394e+00,5.684983e+00,5.092709e+00
1311,Africa Eastern and Southern,AFE,"Service exports (BoP, current US$)",BX.GSR.NFSV.CD,1.105851e+10,1.142424e+10,1.195983e+10,1.653184e+10,1.980731e+10,2.280355e+10,...,4.189568e+10,4.007407e+10,4.444953e+10,4.883753e+10,4.799508e+10,2.972230e+10,3.573699e+10,4.927544e+10,5.332154e+10,5.917083e+10
1426,Africa Eastern and Southern,AFE,"Transport services (% of service exports, BoP)",BX.GSR.TRAN.ZS,2.395546e+01,2.422701e+01,2.372495e+01,2.035498e+01,2.004185e+01,1.975097e+01,...,2.457714e+01,2.246033e+01,2.357277e+01,2.362018e+01,2.556941e+01,3.012808e+01,3.021443e+01,2.944156e+01,2.827607e+01,2.747928e+01
1430,Africa Eastern and Southern,AFE,"Travel services (% of service exports, BoP)",BX.GSR.TRVL.ZS,4.778902e+01,4.923696e+01,5.111144e+01,5.788782e+01,5.704162e+01,5.662932e+01,...,4.670767e+01,4.823710e+01,4.720931e+01,4.689255e+01,4.605773e+01,2.897217e+01,2.863691e+01,3.736879e+01,4.065871e+01,4.178790e+01


In [13]:
df.describe()

,2000,2001,2002,2003,2004,2005,2006,2007,2008,2009,...,2015,2016,2017,2018,2019,2020,2021,2022,2023,2024
count,9.040000e+02,9.210000e+02,9.540000e+02,9.630000e+02,9.760000e+02,1.057000e+03,1.072000e+03,1.084000e+03,1.112000e+03,1.122000e+03,...,1.144000e+03,1.145000e+03,1.130000e+03,1.124000e+03,1.130000e+03,1.125000e+03,1.108000e+03,1.107000e+03,1.105000e+03,1.037000e+03
mean,6.322974e+10,6.048171e+10,6.162018e+10,7.252402e+10,8.775170e+10,9.348213e+10,1.066948e+11,1.238962e+11,1.393422e+11,1.106204e+11,...,1.466591e+11,1.435130e+11,1.612371e+11,1.780854e+11,1.756300e+11,1.604384e+11,2.049245e+11,2.298861e+11,2.275246e+11,2.536897e+11
std,3.929738e+11,3.756094e+11,3.837728e+11,4.505727e+11,5.436594e+11,5.888333e+11,6.725521e+11,7.762684e+11,8.752257e+11,6.836978e+11,...,8.951599e+11,8.736421e+11,9.727358e+11,1.068719e+12,1.045891e+12,9.709820e+11,1.233496e+12,1.367767e+12,1.324994e+12,1.412263e+12
min,-4.300993e-01,-4.278960e-01,-4.285069e-02,-1.566510e-01,8.006886e-04,2.664386e-03,7.673634e-04,4.202122e-03,1.949244e-02,1.422815e-02,...,8.427802e-03,9.122245e-04,1.436626e-02,9.686189e-03,1.339738e-02,-4.443891e-01,2.392616e-02,-2.372141e+01,7.586706e-03,1.239307e-02
25%,1.377051e+01,1.367407e+01,1.395235e+01,1.360820e+01,1.346222e+01,1.318655e+01,1.290857e+01,1.335158e+01,1.258404e+01,1.302538e+01,...,1.153704e+01,1.178157e+01,1.140149e+01,1.148137e+01,1.084875e+01,9.599837e+00,9.334381e+00,1.067821e+01,1.118226e+01,1.107400e+01
50%,4.779966e+01,4.750026e+01,4.808374e+01,4.744169e+01,4.840324e+01,4.806042e+01,4.671560e+01,4.870888e+01,4.564921e+01,4.543644e+01,...,4.405719e+01,4.529032e+01,4.438827e+01,4.282159e+01,4.375895e+01,3.269725e+01,3.605769e+01,3.887247e+01,4.170228e+01,3.960745e+01
75%,1.785472e+09,1.663614e+09,1.765130e+09,1.913869e+09,2.257343e+09,2.329673e+09,2.394869e+09,2.652964e+09,2.674034e+09,2.211979e+09,...,3.216289e+09,3.313035e+09,3.802334e+09,4.370699e+09,4.458396e+09,3.719931e+09,4.598552e+09,5.911379e+09,6.688300e+09,8.630301e+09
max,6.239758e+12,5.965486e+12,6.187966e+12,7.374501e+12,8.991240e+12,1.026518e+13,1.186591e+13,1.371805e+13,1.577955e+13,1.223230e+13,...,1.616521e+13,1.565887e+13,1.737595e+13,1.907189e+13,1.858566e+13,1.724666e+13,2.195042e+13,2.442091e+13,2.321714e+13,2.381458e+13


In [14]:
df_copy=df.copy()

In [15]:
long = df_copy.melt(
    id_vars=[
        'Country Name',
        'Country Code',
        'Indicator Name',
        'Indicator Code'
    ],
    var_name='Year',
    value_name='Value'
)

In [16]:
long['Year'] = long['Year'].astype(int)

In [17]:
long.head()

,Country Name,Country Code,Indicator Name,Indicator Code,Year,Value
0,Africa Eastern and Southern,AFE,"Goods exports (BoP, current US$)",BX.GSR.MRCH.CD,2000,5.769241e+10
1,Africa Eastern and Southern,AFE,Insurance and financial services (% of service...,BX.GSR.INSF.ZS,2000,6.021941e+00
2,Africa Eastern and Southern,AFE,"Service exports (BoP, current US$)",BX.GSR.NFSV.CD,2000,1.105851e+10
3,Africa Eastern and Southern,AFE,"Transport services (% of service exports, BoP)",BX.GSR.TRAN.ZS,2000,2.395546e+01
4,Africa Eastern and Southern,AFE,"Travel services (% of service exports, BoP)",BX.GSR.TRVL.ZS,2000,4.778902e+01


In [18]:
long.drop_duplicates()

,Country Name,Country Code,Indicator Name,Indicator Code,Year,Value
0,Africa Eastern and Southern,AFE,"Goods exports (BoP, current US$)",BX.GSR.MRCH.CD,2000,5.769241e+10
1,Africa Eastern and Southern,AFE,Insurance and financial services (% of service...,BX.GSR.INSF.ZS,2000,6.021941e+00
2,Africa Eastern and Southern,AFE,"Service exports (BoP, current US$)",BX.GSR.NFSV.CD,2000,1.105851e+10
3,Africa Eastern and Southern,AFE,"Transport services (% of service exports, BoP)",BX.GSR.TRAN.ZS,2000,2.395546e+01
4,Africa Eastern and Southern,AFE,"Travel services (% of service exports, BoP)",BX.GSR.TRVL.ZS,2000,4.778902e+01
...,...,...,...,...,...,...
33120,Zimbabwe,ZWE,"Goods exports (BoP, current US$)",BX.GSR.MRCH.CD,2024,7.794595e+09
33121,Zimbabwe,ZWE,Insurance and financial services (% of service...,BX.GSR.INSF.ZS,2024,NaN
33122,Zimbabwe,ZWE,"Service exports (BoP, current US$)",BX.GSR.NFSV.CD,2024,4.571177e+08
33123,Zimbabwe,ZWE,"Transport services (% of service exports, BoP)",BX.GSR.TRAN.ZS,2024,4.180349e+01


In [19]:
long.to_csv('WDI_clean_dataset.csv',index=False)

In [20]:
trade=long.pivot_table(
    index=['Country Name', 'Country Code', 'Year'],
    columns='Indicator Code',
    values= 'Value'
).reset_index()
trade

Indicator Code,Country Name,Country Code,Year,BX.GSR.INSF.ZS,BX.GSR.MRCH.CD,BX.GSR.NFSV.CD,BX.GSR.TRAN.ZS,BX.GSR.TRVL.ZS
0,Afghanistan,AFG,2008,8.204557,5.632992e+08,1.290615e+09,11.465198,1.291595
1,Afghanistan,AFG,2009,8.554064,4.087564e+08,1.701821e+09,6.824291,2.106635
2,Afghanistan,AFG,2010,5.641799,4.357733e+08,2.130238e+09,9.356906,3.520689
3,Afghanistan,AFG,2011,10.288124,4.603184e+08,2.597476e+09,10.368209,3.278467
4,Afghanistan,AFG,2012,8.305200,4.755310e+08,1.853330e+09,13.854985,5.488547
...,...,...,...,...,...,...,...,...
5539,Zimbabwe,ZWE,2020,NaN,4.931846e+09,3.314364e+08,52.619198,18.888094
5540,Zimbabwe,ZWE,2021,NaN,6.359070e+09,2.157347e+08,80.931179,7.497593
5541,Zimbabwe,ZWE,2022,NaN,7.000207e+09,4.532897e+08,34.974239,58.148244
5542,Zimbabwe,ZWE,2023,NaN,7.203513e+09,3.992048e+08,50.949369,38.864766


## PART 1

# Calculate annual growth

How much did services and merchandise exports increase per year

In [21]:
trade = trade.rename(columns={
    'BX.GSR.NFSV.CD': 'Service_Exports',
    'BX.GSR.MRCH.CD': 'Goods_Exports'
})


# 1.Service export growth
trade['Service_Growth_%'] = (
    trade
    .groupby('Country Code')['Service_Exports']
    .pct_change() * 100
)

# 2.Merchandise export growth
trade['Goods_Growth_%'] = (
    trade
    .groupby('Country Code')['Goods_Exports']
    .pct_change() * 100
)

# Calculate total percentage growth

How much did services and merchandise exports increase for 24 years

In [22]:

growth_summary = (
    trade
    .sort_values(['Country Code', 'Year'])
    .groupby('Country Code')
    .agg(
        Service_Start=('Service_Exports', 'first'),
        Service_End=('Service_Exports', 'last'),
        Goods_Start=('Goods_Exports', 'first'),
        Goods_End=('Goods_Exports', 'last')
    )
)

In [23]:
growth_summary['Service_Total_Growth_%'] = (
    (growth_summary['Service_End'] -
     growth_summary['Service_Start'])
    / growth_summary['Service_Start']
) * 100

growth_summary['Goods_Total_Growth_%'] = (
    (growth_summary['Goods_End'] -
     growth_summary['Goods_Start'])
    / growth_summary['Goods_Start']
) * 100

In [24]:
growth_summary

,Service_Start,Service_End,Goods_Start,Goods_End,Service_Total_Growth_%,Goods_Total_Growth_%
Country Code,,,,,,
ABW,1.580927e+09,2.988846e+09,2.346927e+08,1.644509e+08,89.056483,-29.929277
AFE,1.105851e+10,5.917083e+10,5.769241e+10,2.790439e+11,435.070664,383.675227
AFG,1.290615e+09,6.995714e+08,5.632992e+08,7.767345e+08,-45.795490,37.890227
AFW,5.648612e+09,2.540622e+10,3.935678e+10,1.737747e+11,349.778163,341.536826
AGO,2.672800e+08,1.288745e+08,7.920710e+09,3.679504e+10,-51.782980,364.542175
...,...,...,...,...,...,...
XKX,3.097813e+08,3.655679e+09,7.618903e+07,1.034733e+09,1080.083939,1258.112767
YEM,3.720818e+08,4.656326e+08,6.413184e+09,4.728360e+08,25.142515,-92.627126
ZAF,5.045602e+09,1.611476e+10,3.194974e+10,1.115609e+11,219.382368,249.176084


# Calculate CAGR

comparing growth over a long period from 2000 to 2024

In [25]:
n_years = 2024 - 2000

In [26]:
growth_summary['Service_CAGR_%'] = (
    (growth_summary['Service_End'] /
     growth_summary['Service_Start'])
    ** (1 / n_years) - 1
) * 100

In [27]:
growth_summary['Goods_CAGR_%'] = (
    (growth_summary['Goods_End'] /
     growth_summary['Goods_Start'])
    ** (1 / n_years) - 1
) * 100

In [28]:
growth_summary

,Service_Start,Service_End,Goods_Start,Goods_End,Service_Total_Growth_%,Goods_Total_Growth_%,Service_CAGR_%,Goods_CAGR_%
Country Code,,,,,,,,
ABW,1.580927e+09,2.988846e+09,2.346927e+08,1.644509e+08,89.056483,-29.929277,2.689171,-1.471011
AFE,1.105851e+10,5.917083e+10,5.769241e+10,2.790439e+11,435.070664,383.675227,7.238434,6.788153
AFG,1.290615e+09,6.995714e+08,5.632992e+08,7.767345e+08,-45.795490,37.890227,-2.519411,1.347700
AFW,5.648612e+09,2.540622e+10,3.935678e+10,1.737747e+11,349.778163,341.536826,6.465345,6.383340
AGO,2.672800e+08,1.288745e+08,7.920710e+09,3.679504e+10,-51.782980,364.542175,-2.993683,6.608716
...,...,...,...,...,...,...,...,...
XKX,3.097813e+08,3.655679e+09,7.618903e+07,1.034733e+09,1080.083939,1258.112767,10.831456,11.482233
YEM,3.720818e+08,4.656326e+08,6.413184e+09,4.728360e+08,25.142515,-92.627126,0.938893,-10.294680
ZAF,5.045602e+09,1.611476e+10,3.194974e+10,1.115609e+11,219.382368,249.176084,4.957374,5.348136


 # Calculate the Service/Merchandise ratio

In [29]:
trade['Service_Goods_Ratio'] = (
    trade['Service_Exports'] /
    trade['Goods_Exports']
)

In [30]:
trade

Indicator Code,Country Name,Country Code,Year,BX.GSR.INSF.ZS,Goods_Exports,Service_Exports,BX.GSR.TRAN.ZS,BX.GSR.TRVL.ZS,Service_Growth_%,Goods_Growth_%,Service_Goods_Ratio
0,Afghanistan,AFG,2008,8.204557,5.632992e+08,1.290615e+09,11.465198,1.291595,NaN,NaN,2.291171
1,Afghanistan,AFG,2009,8.554064,4.087564e+08,1.701821e+09,6.824291,2.106635,31.861270,-27.435295,4.163411
2,Afghanistan,AFG,2010,5.641799,4.357733e+08,2.130238e+09,9.356906,3.520689,25.174028,6.609547,4.888408
3,Afghanistan,AFG,2011,10.288124,4.603184e+08,2.597476e+09,10.368209,3.278467,21.933594,5.632530,5.642780
4,Afghanistan,AFG,2012,8.305200,4.755310e+08,1.853330e+09,13.854985,5.488547,-28.648798,3.304794,3.897391
...,...,...,...,...,...,...,...,...,...,...,...
5539,Zimbabwe,ZWE,2020,NaN,4.931846e+09,3.314364e+08,52.619198,18.888094,-45.058047,5.750183,0.067203
5540,Zimbabwe,ZWE,2021,NaN,6.359070e+09,2.157347e+08,80.931179,7.497593,-34.909180,28.938927,0.033926
5541,Zimbabwe,ZWE,2022,NaN,7.000207e+09,4.532897e+08,34.974239,58.148244,110.114455,10.082256,0.064754
5542,Zimbabwe,ZWE,2023,NaN,7.203513e+09,3.992048e+08,50.949369,38.864766,-11.931647,2.904281,0.055418


In [31]:
global_trade = (
    trade
    .groupby('Year')[['Service_Exports', 'Goods_Exports']]
    .sum(min_count=1)
    .reset_index()
)

In [32]:
global_trade.head(25)

Indicator Code,Year,Service_Exports,Goods_Exports
0,2000,1.238072e+13,4.477896e+13
1,2001,1.246929e+13,4.323437e+13
2,2002,1.353549e+13,4.525016e+13
3,2003,1.484083e+13,5.499981e+13
4,2004,1.813206e+13,6.751360e+13
5,2005,2.015741e+13,7.865320e+13
6,2006,2.274185e+13,9.163498e+13
7,2007,2.734827e+13,1.069552e+14
8,2008,3.106492e+13,1.238837e+14
9,2009,2.781606e+13,9.630008e+13


# Calculate global growth

In [33]:
global_trade['Service_Growth_%'] = (
    global_trade['Service_Exports']
    .pct_change() * 100
)

global_trade['Goods_Growth_%'] = (
    global_trade['Goods_Exports']
    .pct_change() * 100
)

In [34]:
global_trade['Service_Goods_Ratio'] = (
    global_trade['Service_Exports'] /
    global_trade['Goods_Exports']
)

# Calculate global CAGR

In [35]:
service_cagr = (
    (global_trade['Service_Exports'].iloc[-1] /
     global_trade['Service_Exports'].iloc[0])
    ** (1 / 24) - 1
) * 100

In [36]:
goods_cagr = (
    (global_trade['Goods_Exports'].iloc[-1] /
     global_trade['Goods_Exports'].iloc[0])
    ** (1 / 24) - 1
) * 100

In [37]:
print(f"Service CAGR: {service_cagr:.2f}%")
print(f"Goods CAGR: {goods_cagr:.2f}%")

Service CAGR: 7.30%
Goods CAGR: 6.34%


Between 2000 and 2024, global service exports grew at an estimated compound annual rate of 7.30%, compared with 6.34% for merchandise exports, indicating faster expansion of service exports over the period.

In [38]:
import plotly.express as px

fig = px.line(
    global_trade,
    x='Year',
    y=['Service_Exports', 'Goods_Exports'],
    title='Global Service vs Merchandise Exports (2000–2024)',
    labels={
        'value': 'Export Value (Current US$)',
        'variable': 'Trade Type'
    }
)

fig.show()

In [39]:
fig = px.line(
    global_trade,
    x='Year',
    y=['Service_Growth_%', 'Goods_Growth_%'],
    title='Annual Growth in Global Service vs Merchandise Exports',
    labels={
        'value': 'Annual Growth (%)',
        'variable': 'Trade Type'
    }
)

fig.show()

In [40]:
fig = px.line(
    global_trade,
    x='Year',
    y='Service_Goods_Ratio',
    title='Service-to-Merchandise Export Ratio',
    labels={
        'Service_Goods_Ratio': 'Services / Merchandise'
    }
)

fig.show()

PART 1

Question 1

Did service exports grow faster than merchandise exports?

→ Compare CAGR.

Question 2

How did their annual growth patterns differ?

→ Compare annual percentage growth.

Question 3

Did services become relatively more important?

→ Examine the Service/Merchandise ratio.

Question 4

How did the relationship change over time?

→ Analyze the long-term trend.

Question 5

Were there unusual periods?

→ Investigate years with unusually high/low growth, especially around 2008–09 and 2020–21.

# PART 2

In [41]:
dfcopy2=df.copy()

In [42]:
composition_long = dfcopy2.melt(
    id_vars=[
        'Country Name',
        'Country Code',
        'Indicator Name',
        'Indicator Code'
    ],
    var_name='Year',
    value_name='Percentage'
)

In [43]:
composition_long['Year'] = composition_long['Year'].astype(int)

In [44]:
composition_long.head()

,Country Name,Country Code,Indicator Name,Indicator Code,Year,Percentage
0,Africa Eastern and Southern,AFE,"Goods exports (BoP, current US$)",BX.GSR.MRCH.CD,2000,5.769241e+10
1,Africa Eastern and Southern,AFE,Insurance and financial services (% of service...,BX.GSR.INSF.ZS,2000,6.021941e+00
2,Africa Eastern and Southern,AFE,"Service exports (BoP, current US$)",BX.GSR.NFSV.CD,2000,1.105851e+10
3,Africa Eastern and Southern,AFE,"Transport services (% of service exports, BoP)",BX.GSR.TRAN.ZS,2000,2.395546e+01
4,Africa Eastern and Southern,AFE,"Travel services (% of service exports, BoP)",BX.GSR.TRVL.ZS,2000,4.778902e+01


In [45]:
composition_long['Service_Type'] = (
    composition_long['Indicator Code']
    .map({
        'BX.GSR.TRAN.ZS': 'Transport',
        'BX.GSR.TRVL.ZS': 'Travel',
        'BX.GSR.INSF.ZS': 'Insurance & Financial'
    })
)

In [46]:
composition_long.head()

,Country Name,Country Code,Indicator Name,Indicator Code,Year,Percentage,Service_Type
0,Africa Eastern and Southern,AFE,"Goods exports (BoP, current US$)",BX.GSR.MRCH.CD,2000,5.769241e+10,NaN
1,Africa Eastern and Southern,AFE,Insurance and financial services (% of service...,BX.GSR.INSF.ZS,2000,6.021941e+00,Insurance & Financial
2,Africa Eastern and Southern,AFE,"Service exports (BoP, current US$)",BX.GSR.NFSV.CD,2000,1.105851e+10,NaN
3,Africa Eastern and Southern,AFE,"Transport services (% of service exports, BoP)",BX.GSR.TRAN.ZS,2000,2.395546e+01,Transport
4,Africa Eastern and Southern,AFE,"Travel services (% of service exports, BoP)",BX.GSR.TRVL.ZS,2000,4.778902e+01,Travel


In [47]:
missing = (
    composition_long
    .groupby('Service_Type')['Percentage']
    .apply(lambda x: x.isna().mean() * 100)
    .reset_index(name='Missing_%')
)

missing

,Service_Type,Missing_%
0,Insurance & Financial,22.913208
1,Transport,18.490566
2,Travel,18.400000


Because these indicators are already percentages of service exports, we shouldn't sum them across countries.

Instead, we can calculate a global weighted share if we have the service-export dollar values from Part 1.

In [48]:
trade

Indicator Code,Country Name,Country Code,Year,BX.GSR.INSF.ZS,Goods_Exports,Service_Exports,BX.GSR.TRAN.ZS,BX.GSR.TRVL.ZS,Service_Growth_%,Goods_Growth_%,Service_Goods_Ratio
0,Afghanistan,AFG,2008,8.204557,5.632992e+08,1.290615e+09,11.465198,1.291595,NaN,NaN,2.291171
1,Afghanistan,AFG,2009,8.554064,4.087564e+08,1.701821e+09,6.824291,2.106635,31.861270,-27.435295,4.163411
2,Afghanistan,AFG,2010,5.641799,4.357733e+08,2.130238e+09,9.356906,3.520689,25.174028,6.609547,4.888408
3,Afghanistan,AFG,2011,10.288124,4.603184e+08,2.597476e+09,10.368209,3.278467,21.933594,5.632530,5.642780
4,Afghanistan,AFG,2012,8.305200,4.755310e+08,1.853330e+09,13.854985,5.488547,-28.648798,3.304794,3.897391
...,...,...,...,...,...,...,...,...,...,...,...
5539,Zimbabwe,ZWE,2020,NaN,4.931846e+09,3.314364e+08,52.619198,18.888094,-45.058047,5.750183,0.067203
5540,Zimbabwe,ZWE,2021,NaN,6.359070e+09,2.157347e+08,80.931179,7.497593,-34.909180,28.938927,0.033926
5541,Zimbabwe,ZWE,2022,NaN,7.000207e+09,4.532897e+08,34.974239,58.148244,110.114455,10.082256,0.064754
5542,Zimbabwe,ZWE,2023,NaN,7.203513e+09,3.992048e+08,50.949369,38.864766,-11.931647,2.904281,0.055418


In [49]:
composition = composition_long.merge(
    trade[
        ['Country Code', 'Year', 'Service_Exports']
    ],
    on=['Country Code', 'Year'],
    how='left'
)

###  Calculate the actual dollar value of each service category

In [50]:
composition['Service_Value'] = (
    composition['Service_Exports'] *
    composition['Percentage'] / 100
)

### Calculate global totals

In [51]:

global_composition = (
    composition
    .groupby(['Year', 'Service_Type'])['Service_Value']
    .sum(min_count=1)
    .reset_index()
)

### Calculate each category's global share

In [52]:
# Calculate each category's global share
global_composition['Total_Three_Categories'] = (
    global_composition
    .groupby('Year')['Service_Value']
    .transform('sum')
)

In [53]:
global_composition['Category_Share_%'] = (
    global_composition['Service_Value'] /
    global_composition['Total_Three_Categories']
) * 100

### Find the dominant service category

In [54]:

dominant_service = (
    global_composition
    .sort_values(
        ['Year', 'Service_Value'],
        ascending=[True, False]
    )
    .groupby('Year')
    .first()
    .reset_index()
)

dominant_service[
    ['Year', 'Service_Type', 'Service_Value']
]

,Year,Service_Type,Service_Value
0,2000,Travel,3.769971e+12
1,2001,Travel,3.717061e+12
2,2002,Travel,3.898813e+12
3,2003,Travel,4.341408e+12
4,2004,Travel,5.168951e+12
5,2005,Travel,5.619446e+12
6,2006,Travel,6.137611e+12
7,2007,Travel,7.110445e+12
8,2008,Travel,7.841612e+12
9,2009,Travel,7.156095e+12


In [55]:


fig = px.line(
    global_composition,
    x='Year',
    y='Category_Share_%',
    color='Service_Type',
    title='Composition of Global Service Exports',
    labels={
        'Category_Share_%': 'Share (%)',
        'Service_Type': 'Service Category'
    }
)

fig.show()

In [56]:
fig = px.area(
    global_composition,
    x='Year',
    y='Category_Share_%',
    color='Service_Type',
    title='Changing Composition of Global Service Exports',
    labels={
        'Category_Share_%': 'Share of Selected Services (%)',
        'Service_Type': 'Service Category'
    },
    groupnorm='percent'
)

fig.show()

This lets you visually see whether:

* Travel became more important
* Transport declined
* Financial services increased
* The composition changed substantially around 2020

Analyze COVID-19 separately

In [57]:
periods = pd.cut(
    global_composition['Year'],
    bins=[2014, 2019, 2021, 2024],
    labels=[
        'Pre-COVID',
        'COVID',
        'Post-COVID'
    ]
)

global_composition['Period'] = periods

In [58]:
period_summary = (
    global_composition
    .groupby(['Period', 'Service_Type'], observed=True)
    ['Category_Share_%']
    .mean()
    .reset_index()
)

period_summary

,Period,Service_Type,Category_Share_%
0,Pre-COVID,Insurance & Financial,19.475628
1,Pre-COVID,Transport,32.775683
2,Pre-COVID,Travel,47.748689
3,COVID,Insurance & Financial,29.559461
4,COVID,Transport,43.673538
5,COVID,Travel,26.767002
6,Post-COVID,Insurance & Financial,21.695700
7,Post-COVID,Transport,38.027235
8,Post-COVID,Travel,40.277065


In [59]:
global_composition

,Year,Service_Type,Service_Value,Total_Three_Categories,Category_Share_%,Period
0,2000,Insurance & Financial,1.091067e+12,7.397089e+12,14.749947,NaN
1,2000,Transport,2.536051e+12,7.397089e+12,34.284445,NaN
2,2000,Travel,3.769971e+12,7.397089e+12,50.965608,NaN
3,2001,Insurance & Financial,1.053779e+12,7.323871e+12,14.388275,NaN
4,2001,Transport,2.553031e+12,7.323871e+12,34.859039,NaN
...,...,...,...,...,...,...
70,2023,Transport,1.033540e+13,2.874691e+13,35.953092,Post-COVID
71,2023,Travel,1.219227e+13,2.874691e+13,42.412459,Post-COVID
72,2024,Insurance & Financial,6.794570e+12,3.163143e+13,21.480435,Post-COVID
73,2024,Transport,1.106921e+13,3.163143e+13,34.994336,Post-COVID


 Part 2 

At the end, you'll be able to answer:

Question 1

Which of the three service categories is largest?

Question 2

Has the composition changed from 2000–2024?

Question 3

Which category grew fastest?

Question 4

Did COVID-19 significantly alter service composition?

Question 5

Did travel recover after 2020?

# PART 3

In [60]:
trade['Country Name'].unique()

<StringArray>
[                'Afghanistan', 'Africa Eastern and Southern',
  'Africa Western and Central',                     'Albania',
                     'Algeria',                     'Andorra',
                      'Angola',         'Antigua and Barbuda',
                  'Arab World',                   'Argentina',
 ...
                     'Uruguay',                  'Uzbekistan',
                     'Vanuatu',               'Venezuela, RB',
                    'Viet Nam',          'West Bank and Gaza',
                       'World',                 'Yemen, Rep.',
                      'Zambia',                    'Zimbabwe']
Length: 244, dtype: str

In [61]:
trade[['Country Name', 'Country Code']].drop_duplicates().sort_values('Country Name')

Indicator Code,Country Name,Country Code
0,Afghanistan,AFG
13,Africa Eastern and Southern,AFE
38,Africa Western and Central,AFW
63,Albania,ALB
88,Algeria,DZA
...,...,...
5441,West Bank and Gaza,PSE
5466,World,WLD
5491,"Yemen, Rep.",YEM
5503,Zambia,ZMB


In [62]:
trade[['Country Name', 'Country Code']].drop_duplicates().sort_values('Country Name')


Indicator Code,Country Name,Country Code
0,Afghanistan,AFG
13,Africa Eastern and Southern,AFE
38,Africa Western and Central,AFW
63,Albania,ALB
88,Algeria,DZA
...,...,...
5441,West Bank and Gaza,PSE
5466,World,WLD
5491,"Yemen, Rep.",YEM
5503,Zambia,ZMB


In [63]:
regions = [
    'EAS',   # East Asia & Pacific
    'ECS',   # Europe & Central Asia
    'LCN',   # Latin America & Caribbean
    'MEA',   # Middle East, North Africa, Afghanistan & Pakistan
    'NAC',   # North America
    'SAS',   # South Asia
    'SSF',   # Sub-Saharan Africa
    'AFE',  # Africa Eastern and Southern
    'AFW',  # Africa Western and Central
    'CEB',  # Central Europe and the Baltics
]

In [64]:
regional_trade = trade[
    trade['Country Code'].isin(regions)
].copy()

### Calculate annual growth by region

In [65]:
regional_trade = regional_trade.sort_values(
    ['Country Code', 'Year']
)

regional_trade['Service_Growth_%'] = (
    regional_trade
    .groupby('Country Code')['Service_Exports']
    .pct_change() * 100
)

### Calculate total growth from 2000–2024

In [66]:
regional_growth = (
    regional_trade
    .sort_values(['Country Code', 'Year'])
    .groupby('Country Code')
    .agg(
        Start=('Service_Exports', 'first'),
        End=('Service_Exports', 'last')
    )
    .reset_index()
)

In [67]:
regional_growth['Total_Growth_%'] = (
    (regional_growth['End'] - regional_growth['Start'])
    / regional_growth['Start']
) * 100

In [68]:
regional_growth.sort_values(
    'Total_Growth_%',
    ascending=False
)

,Country Code,Start,End,Total_Growth_%
8,SAS,1.930756e+10,3.965260e+11,1953.734483
2,CEB,4.405128e+10,3.601756e+11,717.628019
4,ECS,8.357546e+11,4.479513e+12,435.984169
0,AFE,1.105851e+10,5.917083e+10,435.070664
3,EAS,3.274887e+11,1.694267e+12,417.351341
9,SSF,1.664871e+10,8.551789e+10,413.660880
5,LCN,6.655930e+10,3.057074e+11,359.300860
1,AFW,5.648612e+09,2.540622e+10,349.778163
7,NAC,3.393029e+11,1.323691e+12,290.120555
6,MEA,NaN,NaN,NaN


### Calculate regional CAGR

In [69]:
regional_growth['CAGR_%'] = (
    (regional_growth['End'] / regional_growth['Start'])
    ** (1 / 24) - 1
) * 100

In [70]:
regional_growth.sort_values(
    'CAGR_%',
    ascending=False
)

,Country Code,Start,End,Total_Growth_%,CAGR_%
8,SAS,1.930756e+10,3.965260e+11,1953.734483,13.419922
2,CEB,4.405128e+10,3.601756e+11,717.628019,9.149853
4,ECS,8.357546e+11,4.479513e+12,435.984169,7.246056
0,AFE,1.105851e+10,5.917083e+10,435.070664,7.238434
3,EAS,3.274887e+11,1.694267e+12,417.351341,7.088064
9,SSF,1.664871e+10,8.551789e+10,413.660880,7.056125
5,LCN,6.655930e+10,3.057074e+11,359.300860,6.558325
1,AFW,5.648612e+09,2.540622e+10,349.778163,6.465345
7,NAC,3.393029e+11,1.323691e+12,290.120555,5.835968
6,MEA,NaN,NaN,NaN,NaN


In [71]:
import plotly.express as px

fig = px.bar(
    regional_growth.sort_values('CAGR_%'),
    x='CAGR_%',
    y='Country Code',
    orientation='h',
    title='Service Export CAGR by Region, 2000–2024',
    labels={
        'CAGR_%': 'CAGR (%)',
        'Country Code': 'Region'
    }
)

fig.show()

In [72]:
regional_trade['Goods_Growth_%'] = (
    regional_trade
    .groupby('Country Code')['Goods_Exports']
    .pct_change() * 100
)

In [73]:
regional_cagr = (
    regional_trade
    .sort_values(['Country Code', 'Year'])
    .groupby('Country Code')
    .agg(
        Service_Start=('Service_Exports', 'first'),
        Service_End=('Service_Exports', 'last'),
        Goods_Start=('Goods_Exports', 'first'),
        Goods_End=('Goods_Exports', 'last')
    )
    .reset_index()
)

In [74]:
regional_cagr['Service_CAGR_%'] = (
    (regional_cagr['Service_End'] /
     regional_cagr['Service_Start'])
    ** (1 / 24) - 1
) * 100

regional_cagr['Goods_CAGR_%'] = (
    (regional_cagr['Goods_End'] /
     regional_cagr['Goods_Start'])
    ** (1 / 24) - 1
) * 100

In [75]:
regional_cagr['CAGR_Difference'] = (
    regional_cagr['Service_CAGR_%'] -regional_cagr['Goods_CAGR_%']
)

In [76]:
regional_cagr

,Country Code,Service_Start,Service_End,Goods_Start,Goods_End,Service_CAGR_%,Goods_CAGR_%,CAGR_Difference
0,AFE,1.105851e+10,5.917083e+10,5.769241e+10,2.790439e+11,7.238434,6.788153,0.450281
1,AFW,5.648612e+09,2.540622e+10,3.935678e+10,1.737747e+11,6.465345,6.383340,0.082005
2,CEB,4.405128e+10,3.601756e+11,1.251656e+11,1.080240e+12,9.149853,9.396007,-0.246154
3,EAS,3.274887e+11,1.694267e+12,1.592150e+12,8.338878e+12,7.088064,7.142923,-0.054860
4,ECS,8.357546e+11,4.479513e+12,2.597576e+12,8.720973e+12,7.246056,5.175970,2.070086
5,LCN,6.655930e+10,3.057074e+11,3.889984e+11,1.624593e+12,6.558325,6.136932,0.421393
6,MEA,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7,NAC,3.393029e+11,1.323691e+12,1.073006e+12,2.647540e+12,5.835968,3.834899,2.001068
8,SAS,1.930756e+10,3.965260e+11,5.609150e+10,5.057336e+11,13.419922,9.595493,3.824429
9,SSF,1.664871e+10,8.551789e+10,9.690751e+10,4.508895e+11,7.056125,6.615749,0.440376


Findings: SAS at 13.42% CAGR is the leading,followed by CEB at 9.15%, then NAC at 5.84%.

The COVID analysis is also strong: travel's selected-category share fell from 47.75% pre-COVID to 26.77% during COVID, while transport increased from 32.78% to 43.67%. Post-COVID, travel recovered to 40.28%.